src/Groups/group_constructors.jl

```diff
@@ -505,34 +525,138 @@ Free group of rank 2
-   @req n >= 0 "n must be a non-negative integer"
-   t = s isa Char ? string(s) : s
+function free_group(L::Vector{<:Symbol}; eltype::Symbol = :letter)
+   @req allunique(L) "generator names must be unique"
+   J = GapObj(L, recursive = true)
    if eltype == :syllable
-     G = FPGroup(GAP.Globals.FreeGroup(n, GapObj(t); FreeGroupFamilyType = GapObj("syllable"))::GapObj)
+     G = FPGroup(GAP.Globals.FreeGroup(J; FreeGroupFamilyType = GapObj("syllable"))::GapObj)
+   elseif eltype == :letter
+     G = FPGroup(GAP.Globals.FreeGroup(J)::GapObj)
    else
-     G = FPGroup(GAP.Globals.FreeGroup(n, GapObj(t))::GapObj)
+     error("eltype must be :letter or :syllable, not ", eltype)
    end
-   GAP.Globals.SetRankOfFreeGroup(GapObj(G), n)
+   GAP.Globals.SetRankOfFreeGroup(GapObj(G), length(J))
    return G
 end
 
-function free_group(L::Vector{<:VarName})
-   J = GapObj(L; recursive = true)
-   G = FPGroup(GAP.Globals.FreeGroup(J)::GapObj)
-   GAP.Globals.SetRankOfFreeGroup(GapObj(G), length(J))
-   return G
+# HACK: we want to use `AbstractAlgebra.@varnames_interface` for free groups,
+# but by default this requires the "constructor" function to have two return
+# values: the ring/group/whatever, and its generator. But `free_group` has
+# traditionally just one return value and we can't change that without badly
+# breaking backwards compatibility. (Besides, I don't really *want* to change
+# this, but that's a different matter).
+#
+# So we use a trick: we introduce `_free_group` with the right return value,
+# and then delegate `free_group` to it, and similarly define the macro
+# `@free_group` by delegating to the `@_free_group` macros (plus some extra
+# shenigans).
+function _free_group(L::Vector{<:Symbol}; eltype::Symbol = :letter)
+   G = free_group(L; eltype)
+   return G, gens(G)
 end
 
-function free_group(L::Vector{<:Char})
-   J = GapObj(Symbol.(L); recursive = true)
-   G = FPGroup(GAP.Globals.FreeGroup(J)::GapObj)
-   GAP.Globals.SetRankOfFreeGroup(GapObj(G), length(J))
-   return G
+AbstractAlgebra.@varnames_interface _free_group(s)
+
+free_group(L0::VarName, Ls::VarName...; kw...) = free_group([L0, Ls...]; kw...)
+free_group(args...; kw...) = _free_group(args...; kw...)[1]
+
+# HACK to get the default variable name stem `:f` instead of `:x`
+# but also to insert validation for `n`.
+function free_group(n::Int, s::VarName = :f; kw...)
+   @req n >= 0 "n must be a non-negative integer"
+  _free_group(n, s; kw...)[1]
+end
+
+"""
+    @free_group(args...)
+
+Return the free group obtained from `free_group(args...)` and introduce
+its generators as Julia variables into the current scope.
+
+# Examples
+```jldoctest
+julia> F = @free_group(:a, :b)
+Free group of rank 2
+
+julia> a^2*b*a*b^-2
+a^2*b*a*b^-2
+```
+
+Note that the `varname => vector` syntax for specifying a vector or
+matrix or general array of variables behaves slightly differently
+compared to `free_group`, as the following example demonstrates.
```

Comment:
Why is this different? Is this something you really want to have or is it just very hard to work around?

Reply:
This was back then agreed upon for the behavior of `@varnames_interface` as users from algebraic geometry wanted e.g. `@polynomial_ring(QQ, :x => (1:2,1:2))` to produce variables `x12` etc. not a matrix `x`.


Honestly I am not that happy with it anymore. I would prefer if we had
- `@polynomial_ring(QQ, "x" => (1:2,1:2))` defines in the current scope a 2x2 matrix `x`
- `@polynomial_ring(QQ, "x#" => (1:2,1:2))` defines the current scope four variables `x11,x12,x21,x22` 

Then hopefully usage would still be convenient enough for users from algebraic geometry? I'll ask @wdecker when I am back from vacation.